src/Tutorials/Custom-Grain-Storage.md

```diff
@@ -0,0 +1,232 @@
+
+    return Task.CompletedTask;
+}
+```
+
+## Reading State
+
+To read the state, to ensure uniqueness, we construct the filename using the grain key and the grain type.
+
+```csharp
+public async Task ReadStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+{
+    var fName = grainReference.ToKeyString() + "." + grainType;
+    var path = Path.Combine(_rootDirectory, fName);
+
+    var fileInfo = new FileInfo(path);
+    if (!fileInfo.Exists)
+        return;
+
+    using (var stream = fileInfo.OpenText())
+    {
+        var storedData = await stream.ReadToEndAsync();
+        grainState.State = JsonConvert.DeserializeObject(storedData, _jsonSettings);
+    }
+}
+```
+
+Note that for the deserialization, we use the `_jsonSettings` which was set on the `Init` function. This is important to be able to serialize/deserialize properly the state.
+
+## Writing State
+
+Writing the state is similar to reading the state.
+
+```csharp
+public async Task WriteStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+{
+    var storedData = JsonConvert.SerializeObject(grainState.State, _jsonSettings);
+
+    var fName = grainReference.ToKeyString() + "." + grainType;
+    var path = Path.Combine(_rootDirectory, fName);
+
+    var fileInfo = new FileInfo(path);
+
+    using (var stream = new StreamWriter(fileInfo.Open(FileMode.Create, FileAccess.Write)))
+    {
+        await stream.WriteAsync(storedData);
+    }
+}
+```
+
+Similarly as reading, we use `_jsonSettings` to write the state.
+
+## Clearing State
+
+Clearing the state would be deleting the file if the file exists.
+
+```csharp
+public Task ClearStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+{
+    var fName = grainReference.ToKeyString() + "." + grainType;
+    var path = Path.Combine(_rootDirectory, fName);
+
+    var fileInfo = new FileInfo(path);
+    if (fileInfo.Exists)
+        fileInfo.Delete();
+
+    return Task.CompletedTask;
+}
+```
+
+## Putting it Together
+
+In order to use the storage, we should be able to configure the root directory. For that we will create an options file `FileGrainStorageOptions`:
+
+```csharp
+public class FileGrainStorageOptions
+{
+    public string RootDirectory { get; set; }
+}
+```
+
+After that we will create a factory which will allow us to scope the options setting to the provider name and at the same time create an instance of the `FileGrainStorage` to ease the registration to the service collection.
+
+```csharp
+public static class FileGrainStorageFactory
+{
+    internal static IGrainStorage Create(IServiceProvider services, string name)
+    {
+        IOptionsSnapshot<FileGrainStorageOptions> optionsSnapshot = services.GetRequiredService<IOptionsSnapshot<FileGrainStorageOptions>>();
+        var options = optionsSnapshot.Get(name);
+        return ActivatorUtilities.CreateInstance<FileGrainStorage>(services, name, options.RootDirectory);
```

Comment:
It's an implementation detail, but we tend to partition our data further using the service Id from ClusterOptions.  This allows the same code with the same configuration to be run on different services and not write to the same location.

Reply:
Oh, I didn't see that partition in the AWS DynamoDB persistence and the Azure blob persistence but I just saw it in the Azure Table impl.
I added the `IOptions<ClusterOptions>` as argument of the constructor and added the `.ServiceId` as part of a function constructing the filename. (`GetKeyString`).